In [5]:
import requests

# URL of the file
url = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv'

# Send a GET request to the URL
response = requests.get(url)

# Save the file
with open('laptops.csv', 'wb') as file:
    file.write(response.content)

print("Download complete!")


Download complete!


In [6]:
import pandas as pd

# Read the CSV file
df = pd.read_csv('laptops.csv')

# Display the first few rows
df.head(10)


Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   
5  MSI Crosshair 17 C12VF-264XES Intel Core i7-12...    New     MSI   
6  Lenovo Thinkpad E14 Gen 4 AMD Ryzen 5 5625U/8G...    New  Lenovo   
7  ASUS VivoBook 15 F515JA-EJ2882W Intel Core i7-...    New    Asus   
8  Medion Akoya E15415 Intel Core i5-10210U/8GB/2...    New  Medion   
9  HP Victus 16-d1038ns Intel Core i7-12700H/16GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   
5   Crosshair  Intel Core i7   32     1000          SSD  RTX 4060    17.3   
6    ThinkPad    AMD Ryzen 5    8      256          SSD       NaN    14.0   
7    VivoBook  Intel Core i7    8      512          SSD       NaN    15.6   
8       Akoya  Intel Core i5    8      256          SSD       NaN    15.6   
9      Victus  Intel Core i7   16      512          SSD  RTX 3050    16.1   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01  
5    No      1699.00  
6    No       909.00  
7    No       809.01  
8    No       519.00  
9    No      1149.00

## Q2. Records count
How many records are in the dataset?

- 12
- 1000
- 2160
- 12160

In [7]:
# Method 1: Using shape
num_records_shape = df.shape[0]
print(f'Number of records (using shape): {num_records_shape}')

# Method 2: Using len()
num_records_len = len(df)
print(f'Number of records (using len): {num_records_len}')

Number of records (using shape): 2160
Number of records (using len): 2160


## Q3. Laptop brands
How many laptop brands are presented in the dataset?

- 12
- 27
- 28
- 2160

In [9]:
# Assuming your DataFrame is named df and the column is 'column_name'
laptop_brands = df['Brand'].nunique()
print(f'Number of unique laptop brands in column "Brand": {laptop_brands}')


Number of unique laptop brands in column "Brand": 27


## Q4. Missing values
How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3

In [10]:
# Find columns with missing values
missing_columns = df.isnull().any()

# Count the number of columns with missing values
num_missing_columns = missing_columns.sum()

print(f'Number of columns with missing values: {num_missing_columns}')

Number of columns with missing values: 3


## Q5. Maximum final price
What's the maximum final price of Dell notebooks in the dataset?

- 869
- 3691
- 3849
- 3936

In [56]:
# Selected Columns
selected_columns_df = df[['Brand', 'Final Price']]
#print(selected_columns_df)

# Specify the category you want to filter by
category_to_filter_Brand = 'Dell'
#Final_Price = selected_columns_df['Final Price'].max()

# Filter the DataFrame for the specific values
filtered_df = selected_columns_df[selected_columns_df['Brand'] == category_to_filter_Brand] #& selected_columns_df['Final Price'].max()
print(filtered_df)

max_value = filtered_df['Final Price'].max()
print(max_value)


     Brand  Final Price
161   Dell       599.00
271   Dell       690.15
281   Dell       889.00
319   Dell       789.00
328   Dell       869.00
...    ...          ...
1843  Dell       420.74
1844  Dell       511.59
1845  Dell       549.93
1846  Dell       699.00
1847  Dell      1086.88

[84 rows x 2 columns]
3936.0


## Q6. Median value of Screen
- Find the median value of *Screen* column in the dataset.
- Next, calculate the most frequent value of the same Screen column.
- Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
- Now, calculate the median value of Screen once again.

In [57]:
# Calculate the median of a specific column
median_value = df['Screen'].median()

print("Median value of Screen:", median_value)

Median value of Screen: 15.6


In [60]:
# Calculate the most frequent value (mode) of a specific column
most_frequent_value = df['Screen'].mode()[0]

print("Most frequent value of Screen:", most_frequent_value)

Most frequent value of Screen: 15.6


In [69]:
# Count the number of NaN values in a specific column
na_count = df['Screen'].isna().sum()

print("Number of NaN values in Column Screen:", na_count)

# Use fillna() to replace NaN values with the most frequent value
df['Screen'] = df['Screen'].fillna(most_frequent_value)

print("DataFrame after filling NaN values with the most frequent value:")
print(df['Screen'])

Number of NaN values in Column Screen: 0
DataFrame after filling NaN values with the most frequent value:
0       15.6
1       15.6
2       15.6
3       15.6
4       15.6
        ... 
2155    17.3
2156    17.3
2157    17.3
2158    13.4
2159    13.4
Name: Screen, Length: 2160, dtype: float64


In [70]:
# New Median Value Again
new_median_value = df['Screen'].median()
print("Median value of Screen:", new_median_value)

Median value of Screen: 15.6


## Has it changed?

Hint: refer to existing mode and median functions to complete the task.

- Yes
- No

- No, the median value is still 15.6

## Q7. Sum of weights
1. Select all the "Innjoo" laptops from the dataset.
2. Select only columns RAM, Storage, Screen.
3. Get the underlying NumPy array. Let's call it X.
4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
5. Compute the inverse of XTX.
6. Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
8. What's the sum of all the elements of the result?
Note: You just implemented linear regression. We'll talk about it in the next lesson.

- 0.43
- 45.29
- 45.58
- 91.30

In [72]:
# 1. Filter the rows where the 'Laptop' column contains the word 'Innjoo'
filtered_df = df[df['Laptop'].str.contains('Innjoo', case=False, na=False)]

print("Filtered DataFrame:")
print(filtered_df)

Filtered DataFrame:
                                                 Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  \
1478  Intel Celeron    8      256          SSD  NaN    15.6    No   
1479  Intel Celeron    8      512          SSD  NaN    15.6    No   
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No   
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No   
1482  Intel Celeron    6      128          SSD  NaN    14.1    No   
1483 

In [74]:
# 2. Select specific columns ('Product' and 'Price') from the DataFrame
selected_columns = filtered_df[['Laptop', 'RAM', 'Storage', 'Screen']]

print("DataFrame with selected columns:")
print(selected_columns)

DataFrame with selected columns:
                                                 Laptop  RAM  Storage  Screen
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    8      256    15.6
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    8      512    15.6
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    4       64    14.1
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    6       64    14.1
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    6      128    14.1
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    6      128    14.1


In [77]:
# 3. Get the underlying NumPy array for the entire DataFrame
X = selected_columns.to_numpy()

print("Underlying NumPy array for the entire DataFrame:")
print(X)

Underlying NumPy array for the entire DataFrame:
[['InnJoo Voom Excellence Intel Celeron N4020/8GB/256GB SSD/15.6"' 8 256
  15.6]
 ['InnJoo Voom Excellence Pro Intel Celeron N4020/8GB/512GB SSD/15.6"' 8
  512 15.6]
 ['Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/14.1"' 4 64 14.1]
 ['Innjoo Voom Laptop Max Intel Celeron N3350/6GB/64GB eMMC/14.1"' 6 64
  14.1]
 ['Innjoo Voom Laptop Pro Intel Celeron N3350/6GB/128GB SSD/14.1"' 6 128
  14.1]
 ['Innjoo Voom Pro Intel Celeron N3350/6GB/128GB eMMC/14.1"' 6 128 14.1]]


In [79]:
# 4. Select only the numerical columns for matrix multiplication
X = df[['RAM', 'Storage', 'Screen']].to_numpy()

# Compute the transpose of X
X_T = X.T

# Compute matrix-matrix multiplication between X.T and X
XTX = X_T @ X  # or np.dot(X_T, X)

print("\nMatrix multiplication result (XTX):")
print(XTX)


Matrix multiplication result (XTX):
[[7.23420000e+05 2.56347520e+07 5.14286600e+05]
 [2.56347520e+07 1.04973173e+09 1.99103300e+07]
 [5.14286600e+05 1.99103300e+07 5.00128313e+05]]


In [81]:
# 5. Compute the inverse of XTX
import numpy as np
try:
    XTX_inv = np.linalg.inv(XTX)
    print("Inverse of XTX:")
    print(XTX_inv)
except np.linalg.LinAlgError:
    print("XTX is a singular matrix and cannot be inverted.")

Inverse of XTX:
[[ 1.09915475e-05 -2.20639891e-07 -2.51893871e-06]
 [-2.20639891e-07  8.31868438e-09 -1.04284462e-07]
 [-2.51893871e-06 -1.04284462e-07  8.74134571e-06]]


In [100]:
# 6. Create an array y with the specified values
y = np.array([1100, 1300, 800, 900, 1000, 1100])

print("Array y:")
print(y)

Array y:
[1100 1300  800  900 1000 1100]


In [101]:
# 7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w = (XTX_inv * XTX) * y
print(w)

ValueError: operands could not be broadcast together with shapes (3,3) (6,) 

In [103]:
import sys
print("Python version:", sys.version)


Python version: 3.11.9 | packaged by Anaconda, Inc. | (main, Apr 19 2024, 16:40:41) [MSC v.1916 64 bit (AMD64)]
